src/Orleans/AssemblyLoader/AssemblyLoader.cs

```diff
@@ -104,33 +104,57 @@ internal class AssemblyLoader
+                var dirs = new Dictionary<string, SearchOption>
+                {
+                    {exeRoot, SearchOption.TopDirectoryOnly}
+                };
 
-            AssemblyLoaderPathNameCriterion[] includeCriteria =
-            {
-                AssemblyLoaderCriteria.IncludeFileNames(new[] {assemblyName})
-            };
-            AssemblyLoaderReflectionCriterion[] loadCriteria =
-            {
-                AssemblyLoaderCriteria.LoadTypesAssignableFrom(typeof (T))
-            };
+                AssemblyLoaderPathNameCriterion[] includeCriteria =
+                {
+                    AssemblyLoaderCriteria.IncludeFileNames(new[] {assemblyName})
+                };
+                AssemblyLoaderReflectionCriterion[] loadCriteria =
+                {
+                    AssemblyLoaderCriteria.LoadTypesAssignableFrom(typeof (T))
+                };
 
-            var discoveredAssemblyLocations = LoadAssemblies(dirs, includeCriteria, loadCriteria, logger);
+                var discoveredAssemblyLocations = LoadAssemblies(dirs, includeCriteria, loadCriteria, logger);
 
-            if (discoveredAssemblyLocations.Count == 0)
-                throw new TypeLoadException("Type " + typeof(T).Name + " wasn't found");
+                if (discoveredAssemblyLocations.Count == 0)
+                {
+                    var error = String.Format("Failed LoadAndCreateInstance for type {0} from assembly {1}. " +
+                                              "Make sure this assembly is in your current executing directory and the type is defined there.", 
+                        typeof(T).FullName, assemblyName);
+                    logger.Error(ErrorCode.Loader_LoadAndCreateInstance_Failure_1, error);
+                    alreadyLoggedError = true;
+                    throw new TypeLoadException(error);
+                }
 
-            if (discoveredAssemblyLocations.Count > 1)
-                throw new TypeLoadException("Type " + typeof(T).Name + " was found more than once");
+                if (discoveredAssemblyLocations.Count > 1)
+                {
+                    var error = String.Format("Type {0} was found more than once in assembly {1}.", typeof(T).FullName, assemblyName);
+                    logger.Error(ErrorCode.Loader_LoadAndCreateInstance_Failure_2, error);
+                    alreadyLoggedError = true;
+                    throw new TypeLoadException(error);
+                }
 
-            var foundType =
-                TypeUtils.GetTypes(discoveredAssemblyLocations, type => typeof(T).IsAssignableFrom(type)).First();
+                var foundType =
+                    TypeUtils.GetTypes(discoveredAssemblyLocations, type => typeof (T).IsAssignableFrom(type)).First();
 
-            return (T)Activator.CreateInstance(foundType, true);
+                return (T) Activator.CreateInstance(foundType, true);
+            }
+            catch (Exception exc)
```

Comment:
If you handle OrleansExceptions and other exceptions differently here, there won't be a need for alreadyLoggedError flag.